lib/tpm2_kdfe.c

```diff
@@ -0,0 +1,238 @@
+    */
+
+    hashInput.size = 4; // room for the counter
+    tpm2_util_concat_buffer(&hashInput, (TPM2B *)Z);
+    tpm2_util_concat_buffer(&hashInput, (TPM2B *)&use);
+    tpm2_util_concat_buffer(&hashInput, (TPM2B *)partyU);
+    tpm2_util_concat_buffer(&hashInput, (TPM2B *)partyV);
+
+    digester d = tpm2_openssl_halg_to_digester(hashAlg);
+
+    for(done = 0, counter = 1;
+        done < bytes;
+        done += hashSize, counter++) {
+
+        counterBE = tpm2_util_hton_32(counter);
+        memcpy(hashInput.buffer, &counterBE, 4);
+
+        d(hashInput.buffer, hashInput.size, resultKey->buffer + done);
+    }
+    // truncate the result to the desired size
+    resultKey->size = bytes;
+
+    return rval;
+}
+
+static EC_POINT * tpm2_get_EC_public_key(TPM2B_PUBLIC *public) {
+    EC_POINT *q = NULL;
+    BIGNUM *bn_Qx, *bn_Qy;
+    EC_KEY *key;
+    const EC_GROUP *group;
+
+    TPMS_ECC_PARMS *tpm_ecc   = &public->publicArea.parameters.eccDetail;
+    TPMS_ECC_POINT *tpm_point = &public->publicArea.unique.ecc;
+
+    int nid = tpm2_ossl_curve_to_nid(tpm_ecc->curveID);
+    if (nid < 0) {
+        return NULL;
+    }
+
+    key = EC_KEY_new_by_curve_name(nid);
+    if (!key) {
+        LOG_ERR("Failed to create EC key from nid");
+        return NULL;
+    }
+
+    group = EC_KEY_get0_group(key);
+    if (!group) {
+        LOG_ERR("EC key missing group");
+        goto out;
+    }
+
+    bn_Qx = BN_bin2bn(tpm_point->x.buffer, tpm_point->x.size, NULL);
+    bn_Qy = BN_bin2bn(tpm_point->y.buffer, tpm_point->y.size, NULL);
+    if((bn_Qx == NULL) || (bn_Qy == NULL)) {
+        LOG_ERR("Could not convert EC public key to BN");
+        goto out;
+    }
+    q = EC_POINT_new(group);
+    if(q == NULL) {
+        LOG_ERR("Could not allocate EC_POINT");
+        goto out;
+    }
+
+    if(0 == EC_POINT_set_affine_coordinates_GFp(group, q, bn_Qx, bn_Qy, NULL)) {
+        LOG_ERR("Could not set affine_coordinates");
+        EC_POINT_free(q);
+        q = NULL;
+        goto out;
+    }
+out:
+    if(bn_Qx) BN_free(bn_Qx);
+    if(bn_Qy) BN_free(bn_Qy);
+    if(key) EC_KEY_free(key);
+
+    return q;
+}
+
+
+static bool get_public_key_from_ec_key(EC_KEY *key, TPMS_ECC_POINT *point) {
+    BIGNUM *x = BN_new();
+    BIGNUM *y = BN_new();
+    const EC_POINT *pubkey = EC_KEY_get0_public_key(key);
+    int nbx, nby;
+    bool result = false;
+
+    if((x == NULL) || (y == NULL) || (pubkey == NULL)) {
+        LOG_ERR("Failed to allocate memory to store EC public key.");
+        goto out;
+    }
+
+    EC_POINT_get_affine_coordinates_GFp(EC_KEY_get0_group(key),
+            pubkey, x, y, NULL);
+    nbx =  BN_num_bytes(x);
+    nby =  BN_num_bytes(y);
+    if( (nbx > sizeof(point->x.buffer))||
+            (nby > sizeof(point->y.buffer))) {
+        LOG_ERR("EC public key has too many bits.");
+        goto out;
+    }
+
+    point->x.size = nbx;
+    point->y.size = nby;
+    BN_bn2bin(x, point->x.buffer);
+    BN_bn2bin(y, point->y.buffer);
```

Comment:
what if this returns NULL?

Reply:
BN_bn2bin() returns length of bignum it writes, never returns NULL. 
BN_bn2bin() will not fail if there are enough buffer to write.